### PR TITLE
Np 581 send response header with data origin(Depends on previous PR)

### PR DESCRIPTION
**Pull request by axthosarouris**
The lambda function sends a Content-Location header with the homeurl of CrossRef or Datacite, depending on where it read the information from

### CODE REVIEW

File: config/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0"?>
 <!DOCTYPE module PUBLIC
-        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
-        "https://checkstyle.org/dtds/configuration_1_3.dtd">
+  "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
```
Comment (axthosarouris) [inline config/checkstyle/checkstyle.xml]: Automatic formatting

File: config/pmd/ruleset.xml [hunk 2]
```diff
@@ -18,7 +18,7 @@
 
   <rule ref="category/java/bestpractices.xml">
     <exclude name="SystemPrintln"/>
-	<exclude name="GuardLogStatement"/>
+    <exclude name="GuardLogStatement"/>
```
Comment (axthosarouris) [inline config/pmd/ruleset.xml]: Autoformatting. Did not edit this file.

File: src/main/java/no/unit/nva/doi/CrossRefClient.java [hunk 3]
```diff
@@ -39,15 +39,16 @@ public CrossRefClient(HttpClient httpClient) {
         this.httpClient = httpClient;
     }
 
-    public Optional<String> fetchDataForDoi(String doiIdentifier) throws URISyntaxException {
+    public Optional<FetchResult> fetchDataForDoi(String doiIdentifier) throws URISyntaxException {
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/doi/CrossRefClient.java]: FetchResult contains the json object and the location from where it was fetched (Datacite, Crossref).

File: src/test/java/no/unit/nva/doi/CrossRefClientTest.java [hunk 4]
```diff
@@ -5,33 +5,20 @@
...
-import no.unit.nva.utils.HttpResponseStatus500;
-import no.unit.nva.utils.MockHttpClient;
+import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Test;
 
 public class CrossRefClientTest extends AbstractLambdaTest {
 
-    public static final String DoiString = "10.1007/s00115-004-1822-4";
```
Comment (axthosarouris) [inline src/test/java/no/unit/nva/doi/CrossRefClientTest.java]: Moved to the abstract class

File: src/test/java/no/unit/nva/doi/CrossRefClientTest.java [hunk 5]
```diff
@@ -40,12 +27,6 @@ void before() throws IOException {
         crossRefClient.setLogger(logger);
     }
 
-    private HttpClient mockHttpClientWithNonEmptyResponse() throws IOException {
```
Comment (axthosarouris) [inline src/test/java/no/unit/nva/doi/CrossRefClientTest.java]: moved to the abstract class

File: src/main/java/no/unit/nva/doi/CrossRefClient.java (hunk 3 above)
Comment (janbj) [reply src/main/java/no/unit/nva/doi/CrossRefClient.java]: Maybe write this comment in code instead?

File: src/main/java/no/unit/nva/doi/FetchResult.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package no.unit.nva.doi;
+
+public class FetchResult {
```
Comment (janbj) [inline src/main/java/no/unit/nva/doi/FetchResult.java]: Give this a more descriptive name

File: template.yaml [hunk 7]
```diff
@@ -96,15 +96,15 @@ Resources:
...
+                doi:
                   type: string
                   description: DOI URL
               required:
                 - doiUrl
   NvaDoiFunction:
     Type: AWS::Serverless::Function # More info about Function Resource: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#awsserverlessfunction
     Properties:
-      CodeUri: .
+      CodeUri: ./build/libs/nva-doi-lambda-1.0-SNAPSHOT.jar
```
Comment (janbj) [inline template.yaml]: Why are we doing this?

Comment (axthosarouris) [reply template.yaml]: Sorry. I forgot to change it back. But, how do you test locally?

File: src/main/java/no/unit/nva/doi/CrossRefClient.java (hunk 3 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/CrossRefClient.java]: Added comment to code.

File: src/main/java/no/unit/nva/doi/FetchResult.java (hunk 6 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/FetchResult.java]: Renamed to MetadataAndContentLocation

File: template.yaml (hunk 7 above)
Comment (axthosarouris) [reply template.yaml]: reversed change to the  original value

File: src/main/java/no/unit/nva/doi/CrossRefClient.java [hunk 8]
```diff
@@ -39,15 +39,24 @@ public CrossRefClient(HttpClient httpClient) {
         this.httpClient = httpClient;
     }
 
-    public Optional<String> fetchDataForDoi(String doiIdentifier) throws URISyntaxException {
+    /**
+     * The method returns the object containing the metadata (title,author etc) of the publication with the specific
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/doi/CrossRefClient.java]: Insert space and trailing comma and abbreviation full stop: title, author, etc.

File: src/main/java/no/unit/nva/doi/CrossRefClient.java [hunk 9]
```diff
@@ -39,15 +39,24 @@ public CrossRefClient(HttpClient httpClient) {
...
+     * DOI, and the source where the metadata were acquired.
+     *
+     * @param doiIdentifier a valid doi identifier or URL.
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/doi/CrossRefClient.java]: Do we catch invalid DOIs?

File: src/main/java/no/unit/nva/doi/CrossRefClient.java [hunk 10]
```diff
@@ -39,15 +39,24 @@ public CrossRefClient(HttpClient httpClient) {
...
+     * @return FetchResult contains the json object and the location from where it was fetched (Datacite, Crossref).
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/doi/CrossRefClient.java]: "JSON", also do we need to enumerate the sources, if we actually return a "location", i.e. a URI?

File: src/main/java/no/unit/nva/doi/DataciteClient.java [hunk 11]
```diff
@@ -8,21 +8,22 @@
 
 public class DataciteClient {
 
-    private static final String dataciteBaseUrlString = "https://data.datacite.org";
+    public static final String DATACITE_BASE_URL_STRING = "https://data.datacite.org";
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/doi/DataciteClient.java]: We should pass the URIs in as environment variables

File: src/main/java/no/unit/nva/doi/CrossRefClient.java [hunk 12]
```diff
@@ -39,15 +39,24 @@ public CrossRefClient(HttpClient httpClient) {
...
+     * @throws URISyntaxException when the input cannot be transformed to a valid URI.
+     */
+    public Optional<MetadataAndContentLocation> fetchDataForDoi(String doiIdentifier) throws URISyntaxException {
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/doi/CrossRefClient.java]: Digital Object Identifier-identifier?

File: src/main/java/no/unit/nva/doi/FetchDoiMetadata.java [hunk 13]
```diff
@@ -29,7 +29,7 @@
     public static final String MISSING_ACCEPT_HEADER = "Missing Accept header";
     public static final Pattern DOI_URL_PATTERN =
         Pattern.compile("^https?://(dx\\.)?doi\\.org/(10(?:\\.[0-9]+)+)/(.+)$",
-                        Pattern.CASE_INSENSITIVE);
+            Pattern.CASE_INSENSITIVE);
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/doi/FetchDoiMetadata.java]: Pattern matching DOIs https://www.crossref.org/blog/dois-and-matching-regular-expressions/

Comment (brinxmat) [reply src/main/java/no/unit/nva/doi/FetchDoiMetadata.java]: Not your commit, I know, but I prefer to see regular expressions outside the code. Either on paper in the bottom of a locked filing cabinet stuck in a disused lavatory with a sign on the door saying 'Beware of the Leopard', or failing that at least in a well-named constant.

File: src/main/java/no/unit/nva/doi/FetchDoiMetadata.java [hunk 14]
```diff
@@ -90,19 +90,22 @@ public GatewayResponse handleRequest(Map<String, Object> input, Context context)
             System.out.println(e.getMessage());
             logger.log(e.getMessage());
             return errorGatewayResponse(e.getMessage(), INTERNAL_SERVER_ERROR.getStatusCode());
-
         }
     }
 
+    private Map<String, String> addContentLocationToHeaders(MetadataAndContentLocation doiMetadata) {
```
Comment (janbj) [inline src/main/java/no/unit/nva/doi/FetchDoiMetadata.java]: Would the other way be bette? metadataAndContentLocation.addToHeaders(Map<String,String> headers)

File: src/main/java/no/unit/nva/doi/CrossRefClient.java (hunk 8 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/CrossRefClient.java]: fixed.

File: src/main/java/no/unit/nva/doi/CrossRefClient.java (hunk 9 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/CrossRefClient.java]: We do some basic checks

File: src/main/java/no/unit/nva/doi/DataciteClient.java (hunk 11 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/DataciteClient.java]: I suggest to create an issue for this.

File: src/main/java/no/unit/nva/doi/FetchDoiMetadata.java (hunk 13 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/FetchDoiMetadata.java]: Create an issue for this?

File: src/main/java/no/unit/nva/doi/FetchDoiMetadata.java [hunk 15]
```diff
@@ -79,9 +77,11 @@ public GatewayResponse handleRequest(Map<String, Object> input, Context context)
...
-            String doiMetadata = lookupDoiMetadata(doiLookup.getDoi(), dataciteContentType);
-            return new GatewayResponse(doiMetadata, OK.getStatusCode(),
-                                       dataciteContentType.getContentType());
+            MetadataAndContentLocation doiMetadata = lookupDoiMetadata(doiLookup.getDoi(), dataciteContentType);
+            Map<String, String> contentHeaderMap = addContentLocationToHeaders(doiMetadata);
+            return new GatewayResponse(doiMetadata.getJson(), OK.getStatusCode(),
+                dataciteContentType.getContentType(),
+                contentHeaderMap);
         } catch (IOException e) {
             System.out.println(e.getMessage());
```
[the remark this answers is not in the document]
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/FetchDoiMetadata.java]: I want to use proper logging, System.out will not be enough at some point.

File: src/main/java/no/unit/nva/doi/DataciteClient.java (hunk 11 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/DataciteClient.java]: https://unit.atlassian.net/browse/NP-690

File: src/main/java/no/unit/nva/doi/CrossRefClient.java (hunk 12 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/CrossRefClient.java]: fixed

File: src/main/java/no/unit/nva/doi/FetchDoiMetadata.java (hunk 13 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/FetchDoiMetadata.java]: https://unit.atlassian.net/browse/NP-691

File: src/main/java/no/unit/nva/doi/CrossRefClient.java [hunk 16]
```diff
@@ -56,7 +65,6 @@ public CrossRefClient(HttpClient httpClient) {
             logger.log(e.getMessage());
```
[the remark this answers is not in the document]
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/CrossRefClient.java]: done

File: src/main/java/no/unit/nva/doi/FetchDoiMetadata.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/FetchDoiMetadata.java]: removed the logger for now

File: src/main/java/no/unit/nva/doi/FetchDoiMetadata.java (hunk 14 above)
Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/FetchDoiMetadata.java]: `addContentLocationToHeaders()` is a pure function. addToHeaders does the same with a side-effect.

Comment (axthosarouris) [reply src/main/java/no/unit/nva/doi/FetchDoiMetadata.java]: renamed and refactored the method

File: src/test/java/no/unit/nva/doi/FetchDoiMetadataTest.java [hunk 17]
```diff
@@ -42,16 +46,22 @@
 
     public static final String VALID_DOI_WITH_DOI_PREFIX = "doi:10.123.4.5/124";
     public static final String VALID_DOI_WITHOUT_DOI_PREFIX = "10.123.4.5/124";
-
+    public static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
```
[the remark this answers is not in the document]
Comment (axthosarouris) [reply src/test/java/no/unit/nva/doi/FetchDoiMetadataTest.java]: fixed

**Pull request merged**